test/e2e/storage/csi_mock/csi_fsgroup_policy.go

```diff
@@ -74,51 +74,115 @@ var _ = utils.SIGDescribe("CSI Mock volume fsgroup policies", func() {
-				}()
-
-				// Ensure that the fsGroup matches what we expect
-				if test.modified {
-					utils.VerifyFSGroupInPod(f, fileName, strconv.FormatInt(*fsGroup, 10), pod)
-				} else {
-					utils.VerifyFSGroupInPod(f, fileName, "root", pod)
+				verifyFSGroupInPod(ctx, m, test.modified)
+
+				// The created resources will be removed by the cleanup() function,
+				// so need to delete anything here.
+			})
+		}
+	})
+
+	ginkgo.Context("CSI FSGroupPolicy Update [LinuxOnly]", func() {
+		tests := []struct {
+			name             string
+			oldFSGroupPolicy storagev1.FSGroupPolicy
+			newFSGroupPolicy storagev1.FSGroupPolicy
+		}{
+			{
+				name:             "should update fsGroup if update from None to File",
+				oldFSGroupPolicy: storagev1.NoneFSGroupPolicy,
+				newFSGroupPolicy: storagev1.FileFSGroupPolicy,
+			},
+			{
+				name:             "should update fsGroup if update from None to default",
+				oldFSGroupPolicy: storagev1.NoneFSGroupPolicy,
+				newFSGroupPolicy: storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy,
+			},
+			{
+				name:             "should not update fsGroup if update from File to None",
+				oldFSGroupPolicy: storagev1.FileFSGroupPolicy,
+				newFSGroupPolicy: storagev1.NoneFSGroupPolicy,
+			},
+			{
+				name:             "should update fsGroup if update from File to default",
+				oldFSGroupPolicy: storagev1.FileFSGroupPolicy,
+				newFSGroupPolicy: storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy,
+			},
+			{
+				name:             "should not update fsGroup if update from detault to None",
+				oldFSGroupPolicy: storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy,
+				newFSGroupPolicy: storagev1.NoneFSGroupPolicy,
+			},
+			{
+				name:             "should update fsGroup if update from detault to File",
+				oldFSGroupPolicy: storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy,
+				newFSGroupPolicy: storagev1.FileFSGroupPolicy,
+			},
+		}
+		for _, t := range tests {
+			test := t
+			ginkgo.It(test.name, func(ctx context.Context) {
+				if framework.NodeOSDistroIs("windows") {
+					e2eskipper.Skipf("FSGroupPolicy is only applied on linux nodes -- skipping")
 				}
+				m.init(ctx, testParameters{
+					disableAttach:  true,
+					registerDriver: true,
+					fsGroupPolicy:  &test.oldFSGroupPolicy,
+				})
+				ginkgo.DeferCleanup(m.cleanup)
+
+				verifyFSGroupInPod(ctx, m, test.oldFSGroupPolicy != storagev1.NoneFSGroupPolicy)
+				m.update(utils.PatchCSIOptions{FSGroupPolicy: &test.newFSGroupPolicy})
+				verifyFSGroupInPod(ctx, m, test.newFSGroupPolicy != storagev1.NoneFSGroupPolicy)
 
 				// The created resources will be removed by the cleanup() function,
 				// so need to delete anything here.
 			})
 		}
 	})
 })
+
+func verifyFSGroupInPod(ctx context.Context, m *mockDriverSetup, modified bool) {
+	// kube-scheduler may need some time before it gets the CSIDriver object.
+	// Without them, scheduling doesn't run as expected by the test.
+	syncDelay := 5 * time.Second
```

Comment:
This may not be enough if run during a load test. It may be wrapping the Pod creation and validation in a retry loop if it encounters errors.

Reply:
Probably need a backoff with the threshold higher than this, or even using exponential backoff if necessary and then give a threshold on when the test should be marked as failed